cmd/snap-update-ns/bootstrap.c

```diff
@@ -0,0 +1,166 @@
+    if (fd < 0) {
+        bootstrap_errno = errno;
+        bootstrap_msg = "cannot open /proc/self/cmdline";
+        return -1;
+    }
+    memset(buf, 0, buf_size);
+    ssize_t num_read = read(fd, buf, buf_size);
+    if (num_read < 0) {
+        bootstrap_errno = errno;
+        bootstrap_msg = "cannot read /proc/self/cmdline";
+    } else if (num_read == buf_size) {
+        bootstrap_errno = 0;
+        bootstrap_msg = "cannot fit all of /proc/self/cmdline, buffer too small";
+        num_read = -1;
+    }
+    close(fd);
+    return num_read;
+}
+
+// find_snap_name scans the command line buffer and looks for the 1st argument.
+const char*
+find_snap_name(char* buf, size_t buf_size, size_t num_read)
+{
+    // cmdline is an array of NUL ('\0') separated strings. We can skip over
+    // the first entry (program name) and look at the second entry, in our case
+    // it should be the snap name.
+    size_t argv0_len = strnlen(buf, buf_size);
+    if (argv0_len + 1 >= num_read) {
+        return NULL;
+    }
+    return &buf[argv0_len + 1];
+}
+
+// setns_into_snap switches mount namespace into that of a given snap.
+static int
+setns_into_snap(const char* snap_name)
+{
+    // Construct the name of the .mnt file to open.
+    char buf[PATH_MAX];
+    int n = snprintf(buf, sizeof buf, "/run/snapd/ns/%s.mnt", snap_name);
+    if (n >= sizeof buf || n < 0) {
+        bootstrap_errno = errno;
+        bootstrap_msg = "cannot format mount namespace file name";
+        return -1;
+    }
+
+    // Open the mount namespace file, note that we don't specify O_NOFOLLOW as
+    // that file is always a special, broken symbolic link.
+    int fd = open(buf, O_RDONLY | O_CLOEXEC);
+    if (fd < 0) {
+        bootstrap_errno = errno;
+        bootstrap_msg = "cannot open mount namespace file";
+        return -1;
+    }
+
+    // Switch to the mount namespace of the given snap.
+    int err = setns(fd, CLONE_NEWNS);
+    if (err < 0) {
+        bootstrap_errno = errno;
+        bootstrap_msg = "cannot switch mount namespace";
+    };
+
+    close(fd);
+    return err;
+}
+
+// sanitize_snap_name performs partial validation of the given name.
+// The goal is to ensure that there are no / or .. in the name.
+int sanitize_snap_name(const char* snap_name)
+{
+    // NOTE: neither set bootstrap_{msg,errno} but the return value means that
+    // bootstrap does nothing. The name is re-validated by golang.
+    if (strstr(snap_name, "..") != NULL) {
+        return -1;
+    }
+    if (strchr(snap_name, '/') != NULL) {
+        return -1;
+    }
+}
+
+// bootstrap prepares snap-update-ns to work in the namespace of the snap given
+// on command line.
+void bootstrap(void)
+{
+// NOTE: This lets us use cgo/go to write tests without running the bulk
+// of the code automatically. In snapd we can just set the required
+// environment variable.
+#define TRIGGER_KEY "SNAPD_INTERNAL"
+#define TRIGGER_VAL "x-switch-namespace=1,"
```

Comment:
Ha, I was just trying to be generic, the`SNAPD_INTERNAL` variable could hold a list of comma-separated key-value pairs. I'm fine with reworking this in any way that we find suitable.